query_builder.go

```diff
@@ -0,0 +1,100 @@
+
+// BuildInsertQuery takes a record, and returns the insert query statement and values representing that record.
+func (q *QueryBuilder) BuildInsertQuery(rec sdk.Record, table string) (string, []interface{}) {
+	keyCols, keyVals, cols, vals := q.getColumnsAndValues(rec.Key.(sdk.StructuredData), rec.Payload.After.(sdk.StructuredData))
+	cols = append(cols, keyCols...)
+	vals = append(vals, keyVals...)
+	query := fmt.Sprintf(insertQuery, table, strings.Join(cols, ", "), q.getPlaceholders(len(cols)))
+	return query, vals
+}
+
+// BuildUpdateQuery takes a record, and returns the update query statement and values representing that record.
+func (q *QueryBuilder) BuildUpdateQuery(rec sdk.Record, table string) (string, []interface{}) {
+	keyCols, keyVals, cols, vals := q.getColumnsAndValues(rec.Key.(sdk.StructuredData), rec.Payload.After.(sdk.StructuredData))
+	setStatement := q.pairValuesWithPlaceholder(cols, setStatementSeparator)
+	whereStatement := q.pairValuesWithPlaceholder(keyCols, whereStatementSeparator)
+	vals = append(vals, keyVals...)
+	query := fmt.Sprintf(updateQuery, table, setStatement, whereStatement)
+	return query, vals
+}
+
+// BuildDeleteQuery takes a record, and returns the delete query statement and values representing that record.
+func (q *QueryBuilder) BuildDeleteQuery(rec sdk.Record, table string) (string, []interface{}) {
+	keyCols, keyVals, _, _ := q.getColumnsAndValues(rec.Key.(sdk.StructuredData), rec.Payload.After.(sdk.StructuredData))
+	whereStatement := q.pairValuesWithPlaceholder(keyCols, whereStatementSeparator)
+	query := fmt.Sprintf(deleteQuery, table, whereStatement)
+	return query, keyVals
+}
+
+// getPlaceholders returns a string of question marks seperated by a comma with a given length.
+func (q *QueryBuilder) getPlaceholders(length int) string {
+	return strings.TrimSuffix(strings.Repeat("?, ", length), ", ")
+}
+
+func (q *QueryBuilder) pairValuesWithPlaceholder(cols []string, separator string) string {
+	if len(cols) == 0 {
+		return ""
+	}
+	return strings.Join(cols, " = ? "+separator+" ") + " = ?"
+}
+
+// getColumnsAndValues returns the key columns and values, and the payload columns and values, each in a slice and in the order mentioned.
+func (q *QueryBuilder) getColumnsAndValues(key, payload sdk.StructuredData) ([]string, []interface{}, []string, []interface{}) {
+	keyColumns := make([]string, 0, len(key))
+	keyValues := make([]interface{}, 0, len(key))
+	columns := make([]string, 0, len(payload))
+	values := make([]interface{}, 0, len(payload))
+
+	// range over both the key and payload values
+	for k, v := range key {
+		keyColumns = append(keyColumns, k)
+		keyValues = append(keyValues, v)
+	}
+
+	// adding commas at the beginning and end of string, to prevent skipping payload fields that have a key as substring.
+	keyColumnsString := "," + strings.Join(keyColumns, ",") + ","
+	for k, v := range payload {
+		// skip Key from payload if exists
+		if strings.Contains(keyColumnsString, ","+k+",") {
```

Comment:
Why don't we check the structured data? Am I missing something? 🤔 
```suggestion
		if _,ok := key[k]; ok {
```

Reply:
I started there and was trying to get the key from the slice for some reason 🥲 my bad